firestore/src/include/firebase/firestore/filter.h

```diff
@@ -0,0 +1,378 @@
+   * equal any of the given values.
+   *
+   * @param[in] field The path of the field to compare.
+   * @param[in] values The list of values to match.
+   *
+   * @return The newly created filter.
+   */
+  static Filter NotIn(const FieldPath& field,
+                      const std::vector<FieldValue>& values);
+
+  /**
+   * @brief Conjunction of a single filter will simply return original filter.
+   *
+   * @param[in] filter Single filter for conjunction.
+   *
+   * @return Same as filter passed in as parameter.
+   */
+  static Filter And(const Filter filter) { return filter; }
+
+  /**
+   * @brief Creates a new filter that is a conjunction of the given filters. A
+   * conjunction filter includes a document if it satisfies all of the given
+   * filters.
+   *
+   * @param[in] filters The filters to perform a conjunction for.
+   *
+   * @return The newly created filter.
+   */
+  template <typename... Filters>
+  static Filter And(const Filters&... filters) {
+    return And(std::vector<Filter>({filters...}));
+  }
+
+  /**
+   * @brief Disjunction of a single filter will simply return original filter.
+   *
+   * @param[in] filter Single filter for disjunction.
+   *
+   * @return Same as filter passed in as parameter.
+   */
+  static Filter Or(const Filter filter) { return filter; }
+
+  /**
+   * @brief Creates a new filter that is a disjunction of the given filters. A
+   * disjunction filter includes a document if it satisfies <em>any</em> of the
+   * given filters.
+   *
+   * @param[in] filters The filters to perform a disjunction for.
+   *
+   * @return The newly created filter.
+   */
+  template <typename... Filters>
+  static Filter Or(const Filters&... filters) {
+    return Or(std::vector<Filter>({filters...}));
+  }
+
+  /**
+   * @brief Copy constructor.
+   *
+   * `Filter` is immutable and can be efficiently copied.
+   *
+   * @param[in] other `Filter` to copy from.
+   */
+  Filter(const Filter& other);
+
+  /**
+   * @brief Move constructor.
+   *
+   * @param[in] other `Filter` to move data from.
+   */
+  Filter(Filter&& other) noexcept;
+
+  /**
+   * @brief Copy assignment operator.
+   *
+   * `Filter` is immutable and can be efficiently copied.
+   *
+   * @param[in] other `Filter` to copy from.
+   *
+   * @return Reference to the destination `Filter`.
+   */
+  Filter& operator=(const Filter& other);
+
+  /**
+   * @brief Move assignment operator.
+   *
+   * @param[in] other `Filter` to move data from.
+   *
+   * @return Reference to the destination `Filter`.
+   */
+  Filter& operator=(Filter&& other) noexcept;
+
+  ~Filter();
+
+ private:
+  friend class Query;
+  friend class QueryInternal;
+  friend class FilterInternal;
+  friend bool operator==(const Filter& lhs, const Filter& rhs);
+  friend struct ConverterImpl;
+
+  static Filter And(const std::vector<Filter>& filters);
+  static Filter Or(const std::vector<Filter>& filters);
```

Comment:
Denver commented on the API doc that it might be better to take the vector by value (as follows) so that the callers can choose to move the vector.
```cpp
  static Filter And(std::vector<Filter> filters);
  static Filter Or(std::vector<Filter> filters);
  ```
I think that's reasonable.